dot/parachain/bitfield-signing/bitfield_signing.go

```diff
@@ -0,0 +1,224 @@
+}
+
+// NewBitfieldSigning creates a new BitfieldSigning subsystem
+func NewBitfieldSigning(overseerChan chan<- any, ks keystore.Keystore, blockState BlockState) *BitfieldSigning {
+	return &BitfieldSigning{
+		subSystemToOverseer: overseerChan,
+		keystore:            ks,
+		bs:                  blockState,
+	}
+}
+
+// Run starts the BitfieldSigning subsystem
+func (b *BitfieldSigning) Run(ctx context.Context, overseerToSubSystem <-chan any) {
+	for {
+		select {
+		case msg := <-overseerToSubSystem:
+			err := b.processMessage(msg)
+			if err != nil {
+				logger.Errorf("processing message: %s", err.Error())
+			}
+		case <-ctx.Done():
+			if err := ctx.Err(); err != nil {
+				logger.Errorf("ctx error: %s\n", err)
+			}
+			return
+		}
+	}
+}
+
+// processMessage processes messages sent to the BitfieldSigning subsystem
+func (b *BitfieldSigning) processMessage(msg any) error {
+	switch msg := msg.(type) {
+	case parachaintypes.ActiveLeavesUpdateSignal:
+		err := b.ProcessActiveLeavesUpdateSignal(msg)
+		if err != nil {
+			return fmt.Errorf("processing active leaves update signal: %w", err)
+		}
+	case parachaintypes.BlockFinalizedSignal:
+		return b.ProcessBlockFinalizedSignal(msg)
+	default:
+		return fmt.Errorf("%w: %T", parachaintypes.ErrUnknownOverseerMessage, msg)
+	}
+	return nil
+}
+
+// Name returns the name of the subsystem
+func (b *BitfieldSigning) Name() parachaintypes.SubSystemName {
+	return parachaintypes.BitfieldSigning
+}
+
+// ProcessActiveLeavesUpdateSignal processes active leaves update signal
+func (b *BitfieldSigning) ProcessActiveLeavesUpdateSignal(signal parachaintypes.ActiveLeavesUpdateSignal) error {
+	activatedLeaf := signal.Activated
+	if activatedLeaf == nil {
+		return nil
+	}
+
+	relayParent := activatedLeaf.Hash
+	rt, err := b.bs.GetRuntime(relayParent)
+	if err != nil {
+		return fmt.Errorf("getting runtime: %w", err)
+	}
+
+	// get validators info
+	validators, err := rt.ParachainHostValidators()
+	if err != nil {
+		return fmt.Errorf("getting validators: %w", err)
+	}
+	validatorID, validatorIndex := parachainutil.SigningKeyAndIndex(validators, b.keystore)
+	if validatorID == nil {
+		// skip the logic if not a validator node
+		return nil
+	}
+
+	// wait for availability distribution has the chance to make candidates available.
+	time.Sleep(availabilityDistributionWaitingPeriod)
+
+	// construct the bitfield according to the availability store
+	bitfield, err := constructAvailabilityBitfield(rt, validatorIndex, b.subSystemToOverseer)
+	if err != nil {
+		return fmt.Errorf("construct availabilityBitfield: %w", err)
+	}
+
+	// signing process
+	data, err := bitfield.MarshalSCALE()
+	if err != nil {
+		return fmt.Errorf("marshal bitfield for signing: %w", err)
+	}
+	validatorPublicKey, err := sr25519.NewPublicKey(validatorID[:])
+	if err != nil {
+		return fmt.Errorf("getting signer's public key: %w", err)
+	}
+	signatureBytes, err := b.keystore.GetKeypair(validatorPublicKey).Sign(data)
```

Comment:
We need to sign bitfield+signingContext, not only bitfield.

[Here](https://github.com/paritytech/polkadot-sdk/blob/934c091421af4be362839996bfaa441ba59bf12b/polkadot/node/core/bitfield-signing/src/lib.rs#L259) is the Rust code to sign bitfield, if you nevigate through nested sign function you will end up to [this](https://github.com/paritytech/polkadot-sdk/blob/934c091421af4be362839996bfaa441ba59bf12b/polkadot/primitives/src/v8/signed.rs#L281) function where bitfield+signingContext signed, not only bitfield.

Reply:
I opened https://github.com/ChainSafe/gossamer/pull/4564 PR to make validator sign method to support any data type.

You can use that method directly to sign bitfield.

Reply:
We already have sign logic implemented but it's in `backing` package. also it only accept `statementVDT` type.
I opened https://github.com/ChainSafe/gossamer/pull/4564 PR to make validator sign method to support any data type.

You can use that method directly to sign bitfield.